rfcs/0044-ckb-light-client/0044-ckb-light-client.md

```diff
@@ -0,0 +1,508 @@
+    - Let $D_{start}$ denote the total difficulty of the start block and
+      $D_{end}$ denote the total difficulty of the tip block.
+    - Calculate the difficulty boundary $D_{boundary}$, which is the start
+      boundary of last $\delta$ fraction region, with formula
+      $D_{boundary} = D_{start} + (1 - \delta)(D_{end} - D_{start})$.
+    - Calculate $\delta = c ^ {k}$.
+    - Use the PDF $g(x)$ to random $m - L$ difficulties, which are satisfied
+      $d_{i} \in \left[ D_{start}, D_{boundary} \right), i \in \left[ 0, m-L-1 \right]$
+      as the samples.
+
+After sampling, the client sends a request to the server, namely, a CKB full
+node, with following data:
+- The blocks hash of the tip block that client knows.
+- The blocks hash of the start block.
+- The blocks number of the start block.
+- The number $L$.
+- The difficulty boundary $D_{boundary}$.
+- The difficulty samples. If there are no difficulties sampled in the
+  previous step, leave them to be empty.
+
+At last, the server replies the proof of the sampled blocks.
+
+Accordingly, the client can treat that tip block as a valid tip block after
+checking the returned proof.
+
+#### Phase #2: Client asks a proof for some blocks.
+
+- At the start, the client send the `GetBlocksProof` message, which contains
+  the last block hash it knows, and hashes of the blocks which require
+  verifying, to the server which has the best chain.
+
+- Then, the server should reply a `SendBlocksProof` message.
+
+  If the last block that the client known is in the server's current chain,
+  then the server should return
+  - the verifiable header of the last block that the client known,
+  - a proof for blocks which are from the provided block hashes and in the
+    current chain,
+  - the headers of those proved blocks,
+  - hashes of blocks which are from the provided block hashes but not in the
+    current chain.
+
+  Otherwise, the server should only return its current verifiable tip
+  header.
+
+#### Phase #3: Client asks a proof for some transactions.
+
+- At the start, the client send the `GetTransactionsProof` message, which
+  contains the last block hash it knows, and hashes of the transactions
+  which require verifying, to the server which has the best chain.
+
+- Then, the server should reply a `SendTransactionsProof` message.
+
+  If the last block that the client known is in the server's current chain,
+  then the server should return
+  - the verifiable header of the last block that the client known,
+  - a proof for blocks which contain valid transactions from the provided
+    transaction hashes,
+  - the proofs that proving the those valid transactions are in those blocks,
+  - hashes of transactions which are from the provided transaction hashes
+    but not in the current chain.
+
+  Otherwise, the server should only return its current verifiable tip
+  header.
+
+## Limitations
+
+To avoid attacks by malicious clients, there are few limitations in server
```

Comment:
We could set some limitations for performance, such as requiring `difficulties` sorted.

Reply:
I will update this section, later.